consensus/src/chained_bft/chained_bft_smr_test.rs

```diff
@@ -819,3 +820,64 @@ fn chain_with_nil_blocks() {
         assert!(nodes[2].smr.block_store().unwrap().root().round() >= 1)
     });
 }
+
+#[test]
+/// Test secondary proposal processing
+fn secondary_proposers() {
+    let runtime = consensus_runtime();
+    let mut playground = NetworkPlayground::new(runtime.executor());
+
+    let nodes = SMRNode::start_num_nodes(3, 2, &mut playground, MultipleOrderedProposers);
+    block_on(async move {
+        // Node 0 is disconnected.
+        playground.drop_message_for(&nodes[0].author, nodes[1].author);
+        playground.drop_message_for(&nodes[0].author, nodes[2].author);
+        // Run a system until node 0 is a designated primary proposer. In this round the
+        // secondary proposal should be voted for and attached to the timeout message.
+        let timeout_msgs = playground
+            .wait_for_messages(2 * 2, NetworkPlayground::timeout_msg_only)
+            .await;
+        let mut secondary_proposal_ids = vec![];
+        for mut msg in timeout_msgs {
+            let timeout_msg = TimeoutMsg::from_proto(msg.1.take_timeout_msg()).unwrap();
+            assert!(timeout_msg.pacemaker_timeout().vote_msg().is_some());
+            secondary_proposal_ids.push(
+                timeout_msg
+                    .pacemaker_timeout()
+                    .vote_msg()
+                    .unwrap()
+                    .proposed_block_id(),
+            );
+        }
+        assert_eq!(secondary_proposal_ids.len(), 4);
+        let secondary_proposal_id = secondary_proposal_ids[0];
+        for id in secondary_proposal_ids {
+            assert_eq!(secondary_proposal_id, id);
+        }
+        // Stop dropping messages (to make counting easier) and run 3 rounds to commit that id.
```

Comment:
can we run 3 secondary round to commit the block?

Reply:
It's hard to write a test that would ensure 3 secondary rounds in a row: the primaries are chosen based on the hashes of rounds, so it's hard to know what rounds are going to stay without primary.